wallet-new/src/Cardano/Wallet/Kernel/DB/Spec/Update.hs

```diff
@@ -285,24 +278,29 @@ observableRollbackUseInTestsOnly = rollback
 -- (to the rollback) and the transactions that got removed from pending
 -- (since they are new confirmed).
 switchToFork :: SecurityParameter
-             -> Int  -- ^ Number of blocks to rollback
+             -> Maybe Core.HeaderHash  -- ^ Roll back until we meet this block.
              -> OldestFirst [] PrefilteredBlock -- ^ Blocks to apply
              -> Update' ApplyBlockFailed
                         (Checkpoints Checkpoint)
                         (Pending, Set Txp.TxId)
-switchToFork k numRollbacks blocksToApply = do
-    reintroduced <- rollbacks Pending.empty numRollbacks
+switchToFork k oldest blocksToApply = do
+    reintroduced <- rollbacks Pending.empty
     applyBlocks reintroduced Set.empty (getOldestFirst blocksToApply)
   where
     rollbacks :: Pending -- Accumulator: reintroduced pending transactions
-              -> Int     -- Number of rollback to do
               -> Update' e
                          (Checkpoints Checkpoint)
                          Pending
-    rollbacks !accNew 0 = return accNew
-    rollbacks !accNew n = do
+    rollbacks !accNew = do
+        curCtx <- use currentContext
         reintroduced <- rollback
-        rollbacks (Pending.union accNew reintroduced) (n - 1)
+        let acc = Pending.union accNew reintroduced
+            prev = join (curCtx <&> _bcPrevMain) <&> _fromDb
+
+        case (prev == oldest, prev) of
+            (True, _)        -> return acc    -- We rolled back everything we needed to.
+            (False, Nothing) -> return acc    -- The checkpoints began after the fork point.
```

Comment:
I think this is is a tricky one. Under normal circumstances, suppose we have a chain

```
A -> B -> C -> D -> E
```

and we want to roll back to `C`. If the account was created at point `D`, it indeed won't have a checkpoint for `C` and rolling back to the "genesis" (of the account) is correct. However, if this accounts happens to live in some kind of fork and we don't find `C` because of that reason, then we have no choice but to start restoration for this account since it's state is basically unknown. 

Maybe what needs to happen is that instead of giving only the final hash, we give every hash that we expect to see along the way, so that if we see a hash that's unexpected, we know that something is wrong with the account and we need to start restoration?

Reply:
> Maybe what needs to happen is that instead of giving only the final hash, we give every hash that we expect to see along the way, so that if we see a hash that's unexpected, we know that something is wrong with the account and we need to start restoration?

This is a little trickier than it seems. Say you had a wallet from the future; whatever list of expected hashes we are passing to `switchToFork` here may not know about the most recent predecessors of the wallet's current checkpoint, so we would immediately see an unexpected hash even if the wallet is on the correct fork.

I think we should make a ticket for this case and think it through more carefully. In the meantime, the go-straight-to-restoration tactic should at least be correct here, if not optimal.